### PR TITLE
Factor out getting header meta as YAML from table into meta.py module

### PR DESCRIPTION
**Pull request by taldcroft**
The io.ascii ECSV reader class had functionality for serializing Table header meta info (table.meta and column meta attributes) as YAML.  It also had code for taking that same serialized YAML string and returning a data structure that can be used for updating the same Table header meta info.  This PR factors out that functionality into a new module `meta.py` within the `table` package.

The driver for this is to allow the FITS Table I/O connector to (somehow) store the same YAML string representing table meta and allow for round-tripping of Table objects with meta through FITS.  The implementation of this is left for @nhmc, probably with help from @embray to decide the best path.

While we're at it, the same can be done for HDF5 @astrofrog.

### CODE REVIEW

**Comment by embray** [comment]
Originally I told @nhmc I'd be against specifically adding anything for storing the metadata in a FITS file.  But on second thought we're already looking to add the ability to stuff an ASDF file into a FITS extension.  And in fact a plain YAML document with no binary blocks can still be a valid ASDF file.

So I'll go back on myself and say I'd be okay with including in Astropy an implementation for a convention of storing YAML in a FITS header--this could in turn be used to serialize (where applicable) the Table.meta to FITS.  However, if we do that I'd like to see at the very least an APE describing this FITS convention.  And preferably also to get this added to the list of registered FITS conventions, though I care slightly less about that.

**Comment by embray** [comment]
In particular, I assume the YAML data would go in a separate extension, and there needs to be a convention defined for associating a YAML extension with a particular TABLE or BINTABLE extension.  The other question is do we define a new extension type (eg. "YAML"?) or do we just use an IMAGE with bitpix=8 and store the raw YAML bytes?

File: astropy/table/meta.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+    lines : list
+        List of text lines with YAML header content
+
+    Returns
+    -------
+    header : dict
+        Dictionary describing table and column meta
+    """
+
+    import textwrap
```
Comment (embray) [inline astropy/table/meta.py]: I think this in-line import could be done away with.

File: astropy/table/meta.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+    try:
+        import yaml
+    except ImportError:
+        raise ImportError('`import yaml` failed, PyYAML package is required for ECSV format')
```
Comment (embray) [inline astropy/table/meta.py]: This reminds me how much I want to revisit #476 and come up with a way to do this.  #3389 might also help with that.  Sorry, I'm getting off topic though...

File: astropy/table/meta.py (hunk 1 above)
Comment (taldcroft) [inline astropy/table/meta.py]: You mean just put it at the top?

File: astropy/table/meta.py (hunk 2 above)
Comment (taldcroft) [inline astropy/table/meta.py]: Note to self: get rid of `for ECSV format` since that might not always be the reader now.

**Comment by taldcroft** [comment]
@embray - I fully agree with your comments about the need for formal process and wider involvement when going forward with the FITS implementation.  Likely the same would apply for HDF5.  This PR is not attempting to get involved in those battles, just laying the foundation...  :smile:

File: astropy/table/meta.py (hunk 1 above)
Comment (embray) [inline astropy/table/meta.py]: Yeah, no reason to inline this.

**Comment by embray** [comment]
@taldcroft Right, I understand that, and for that purpose it's fine.  I'll open a new issue if there isn't one already for the FITS convention.  I'm not sure exactly what's needed for HDF5 though?  I thought we already had a way of round-tripping (most) Astropy tables to HDF5.

**Comment by nhmc** [comment]
Thanks for this @taldcroft.

My motivation is to allow non-trivial metadata (which can be represented as yaml) to be saved in any format. I think we should be able to assign metadata to a table and know it will be saved along with the data.  Currently table metadata is completely lost when saving to an ascii format other than enhanced csv, and only partially saved in FITS and HDF5.  Neither FITS nor HDF5 can handle a dictionary as a metadata entry, for example.   The formats I care most about are fixed width ascii and FITS, because that's what I use.

**Comment by taldcroft** [comment]
As @nhmc mentioned, right now the HDF5 I/O layer has limited support for values in the `table.meta` attribute, but ignores all column meta.  The implementation is by mapping `table.meta` keys to HDF5 attributes.  I noticed in the h5py docs that it is "recommended" that these attributes be less than 64k bytes long.  To me this implies we would be better off moving the HDF5 layer in the same direction as FITS and putting the YAML-serialized meta string into a separate data structure.  This does have a downside of

- Being less natural for "simple" cases of just a few meta attributes that are simple scalars
- Not back compatible

Maybe a keyword arg to choose between meta representation methods?

**Comment by nhmc** [comment]
I think it could be made backwards compatible. The metadata could still be
written as is currently being done in FITS and HDF5. In cases where all the
metadata can be correctly saved in this way, then no YAML-serialised
extension would be written and there is no change from the current
behaviour. If some of the metadata is more complicated, though, then a YAML
extension would be written. This extension would take precedence over any
other keywords when reading the file in again.

**Comment by embray** [comment]
I think it's fine to try to come up with conventions (that are well documented, accepted) where appropriate.  ECSV is fine.  I'm more wary of shoving things into FITS, etc., but not absolutely opposed.  But there are some file formats that simply will not round-trip all information stored in an Astropy Table object (at least not completely).  I think it's important that we're clear on that point.

**Comment by taldcroft** [comment]
I won't stop until we can round trip a Table through GIF.  Why?  Because it's there.

**Comment by taldcroft** [comment]
Does anyone have objections to this PR or need more time to review?

File: astropy/io/ascii/ecsv.py [hunk 3]
```diff
@@ -227,76 +48,29 @@ def write(self, lines):
...
-        header['datatype'] = [_get_col_attributes(col) for col in self.cols]
-
         # Set the delimiter only for the non-default option(s)
         if self.splitter.delimiter != ' ':
             header['delimiter'] = self.splitter.delimiter
 
-        header_yaml = yaml.dump(header, Dumper=TableDumper)
-        outs = ['%ECSV {0}'.format(ECSV_VERSION), '---']
-        outs.extend(header_yaml.splitlines())
+        header_yaml_lines = (['%ECSV {0}'.format(ECSV_VERSION),
```
Comment (mdboom) [inline astropy/io/ascii/ecsv.py]: Minor change recommended here.  I suspect you probably got this from an earlier version of ASDF.  It turns out the YAML standard doesn't strictly like the `%` prefix here, but you can use `#` (comment) instead.

**Comment by mdboom** [comment]
:+1: from me.

File: astropy/io/ascii/ecsv.py (hunk 3 above)
Comment (taldcroft) [inline astropy/io/ascii/ecsv.py]: Bummer.  The `%` prefix is now required in the ECSV 1.0 standard in APE-6.

Regardless, this PR is just a refactoring and shouldn't be changing any functionality (for better or worse).  Are you OK with this @mdboom ?

Comment (mdboom) [inline astropy/io/ascii/ecsv.py]: Yeah -- that's fine to not fix now.

**Pull request merged**